apache/superset
# fix(sqla): make text clause escaping optional

### SUMMARY
A recent PR #17419 enabled global escaping of the colon character in queries, as the majority of SQLAlchemy dialects support escaping the colon character (see #17098 why this is needed). Escaping has been verified to work correctly on Postgres, MySql, MSSQL, Sqlite, BigQuery, Druid, Hive and Redshift, so it appears safe to assume escaping is supported on the majority of dialects. However, PyAthena does not, and adds the escape character to the final query, which breaks any query containing a colon character.

To support literal colons in queries, but fix the regression for Athena, this PR adds a flag `allows_escaped_colons` to `BaseEngineSpec`, which defaults to `True`, and is set to `False` in `PyAthenaSpec`. Relevant unit tests are added to the `BaseEngineSpec` and `AthenaEngineSpec`. In addition, all athena integration tests are converted to unit tests in pytest format.

### BEFORE
Currently adding a virtual table with the query `select '123:456' as test_str` and then adding a Custom SQL query `test_str = '123:456'` generates the following query on Athena (notice the backslash before the colon characters):
```sql
SELECT "test_str" AS "test_str"
FROM
  (select '123\:456' as test_str) AS "virtual_table"
WHERE ((test_str = '123\:456'))
LIMIT 1000
```

For comparison, the same virtual table + filter produces the following on Postgres + all other tested databases:
```sql
SELECT test_str AS test_str
FROM
  (select '123:456' as test_str) AS virtual_table
WHERE ((test_str = '123:456'))
LIMIT 1000
```

### TESTING INSTRUCTIONS
1. create a virtual table `select '123:456' as test_str` referencing an Athena database
2. Create a chart based on the virtual table and add a filter `test_str = '123:456'`
3. Notice how the rendered query has escaped both the subquery and where clause

### ADDITIONAL INFORMATION
